### ISSUE TITLE
Using the iocp reactor, twisted is sending duplicate / corrupt data under load.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9353]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)| @archer0001 reported|
|-|-|
|Trac ID|trac#9353|
|Type|defect|
|Created|2017-12-18 14:33:50Z|

I used twisted to drive tests of a product I help develop.  I've noticed that when I have >4000 connections, the python client starts sending malformed data.

After much debugging, I've tracked the issue down to _cbRead in internet/iocpreactor/abstract.py.

I believe the order of events are

1. tcp.py:write is called with new data.  Data is added to the temp buffer and the write is scheduled.
1. A previous write completes and _cbWrite is called.
1. _handleWrite moves the temp data dataBuffer and doWrite is called.
1. The previously scheduled write is triggered, and calls _cbWrite, which sends the same data since offsets haven't been updated yet.
1. Another send with a different amount of data can be scheduled now, and if the 2nd erroneous write returns, it updates the offsets incorrectly and now a data fragment is sent, in addition to duplicate data.

I am attaching a patch to this that fixes the issue, though I've not yet written tests to exercise it, since I think I'll have to monkeypatch iocpsupport's getEvent to delay getting an event until a second write is scheduled.  I'm unsure how invasive I'm allowed to be in my tests.

Attachments:

* [fix_example.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d26/d2676e5fb29ed985e452aa6434e13d06eb60e1be/888a6893879a67b3051b2dc402d6279358846f79.patch) (536 bytes) - added by archer0001 on 2017-12-18 14:35:51Z - Proposed Fix.  This patch is incomplete since it doesn't include tests.
* [bugfix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d26/d2676e5fb29ed985e452aa6434e13d06eb60e1be/52fc032535bd68d4cfbb32b4352f856be867cea4.patch) (7060 bytes) - added by archer0001 on 2017-12-28 13:51:51Z - Patch of the Pull Request.

<details><summary>Searchable metadata</summary>

```
trac-id__9353 9353
type__defect defect
reporter__archer0001 archer0001
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__None None
time__1513607630708677 1513607630708677
changetime__1568870736300919 1568870736300919
version__None None
owner__archer0001 archer0001

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)<a name="note_1"></a>|@archer0001 commented|
|-|-|

Updated to fix list formatting.

**Comment by twisted-trac** [comment]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)<a name="note_2"></a>|@archer0001 commented|
|-|-|

Ugh, I think the patch I added a naïve fix.  The same problem arises if the IOEvent hasn't completed and more data is written to the tcp port.  Since the offsets aren't updated, duplicate data is sent.

**Comment by twisted-trac** [comment]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)<a name="note_3"></a>|@archer0001 commented|
|-|-|

Unit test exercising the bug found in: https://github.com/archer0001/twisted/tree/9353-archer0001-bugexercise

**Comment by twisted-trac** [comment]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)<a name="note_4"></a>|@archer0001 commented|
|-|-|

PR: https://github.com/twisted/twisted/pull/948

**Comment by twisted-trac** [comment]
|[<img alt="archer0001's avatar" src="https://avatars.githubusercontent.com/u/8215860?s=50" width="50" height="50">](https://github.com/archer0001)<a name="note_5"></a>|@archer0001 commented|
|-|-|

fix_example was the naive fix.  bugfix.patch should contain all the changes to address the issue.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @archer0001|
|-|-|

See https://github.com/twisted/twisted/pull/948#pullrequestreview-102851097

Maybe use the code from [[#9358](https://github.com/twisted/twisted/issues/9358)](https://github.com/twisted/twisted/issues/9358) to trigger a test failure and see if those tests pass after applying the fix from here.

As long as the branch has a nice history, in which I can see the test failing and then another commit with the fix in which the tests pass I am happy to review and merge this and #9358 in the same branch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wiml's avatar" src="https://avatars.githubusercontent.com/u/156891?s=50" width="50" height="50">](https://github.com/wiml)<a name="note_7"></a>|@wiml commented|
|-|-|

I am curious if this is the same as ticket:9446 ? It sounds at least superficially similar.

**Comment by twisted-trac** [comment]
|[<img alt="wiml's avatar" src="https://avatars.githubusercontent.com/u/156891?s=50" width="50" height="50">](https://github.com/wiml)<a name="note_8"></a>|@wiml commented|
|-|-|

This looks like it might also be the same as [#3525](https://github.com/twisted/twisted/issues/3525) , which includes some discussion of how to test for it, and other patches — it might be worth checking to see if it is indeed the same problem.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set status to `closed`|
|-|-|

Pretty sure that this was the issue fixed in [#9446](https://github.com/twisted/twisted/issues/9446).

If anyone has evidence to the contrary, please feel free to comment and reopen.
